### PR TITLE
[FLINK-4202] Add restarting time JM metric

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR adds a JM metric which shows the time it took to restart a job. The time is
measured between entering the JobStatus.RESTARTING and reaching the JobStatus.RUNNING
state. During this time, the restarting time is continuously updated. The metric only
shows the time for the last restart attempt.

R @zentol if you find time.

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphMetricsTest.java [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ExecutionGraphMetricsTest extends TestLogger {
+
+	/**
+	 * This test tests that the restarting time metric correctly displays restarting times.
+	 */
+	@Test
+	public void testExecutionGraphRestartTimeMetric() throws JobException, IOException, InterruptedException {
```
Comment (uce) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphMetricsTest.java]: Very good test 👍

**Comment by uce** [comment]
+1 to merge. Tests and and comments are very good. I've tested this locally with a job with restarts and everything works as expected:

screen shot 2016-07-21 at 15 47 34

**Comment by tillrohrmann** [comment]
Thanks for the review @uce. Will merge this PR.

**Comment by zentol** [comment]
The Javadoc for the `RestartingTimeGauge` is not consistent with the implementation.

The Javadocs say "The time between the RESTARTING and RUNNING".

The implementation says either
* the time between RESTARTING and RUNNING
* the time since RESTARTING
* the time between RESTARTING and FAILED/SOME_OTHER_TERMINAL_STATE

**Comment by zentol** [comment]
Also, the documentation (list of all exposed metrics) was not updated.

**Comment by tillrohrmann** [comment]
Good points with the documentation and the exposed metrics. I'll update the documentation.

**Pull request closed without merging**